lesson28/src/js/password.js

```diff
@@ -0,0 +1,128 @@
+
+const isInvalidRegex = (reg, value) => reg.test(value) ? false : true;
+
+const addValidClassName = (target) => {
+  target.parentElement.classList.add("valid");
+};
+
+const isValidField = (e) => {
+  const field = e.target;
+  if (isBlankInInput(field.value)) {
+    addInvalidMessage(field, "未入力です");
+    submitButton.disabled = true;
+    return;
+  }
+  if (constraint[field.id].validation()) {
+    addInvalidMessage(field, constraint[field.id].invalidMessage);
+    submitButton.disabled = true;
+    return;
+  }
+
+  addValidClassName(field);
+  toggleDisableSubmitButton();
+};
+
+const addInvalidMessage = (target, message) => {
+  const parent = target.parentElement;
+  parent.classList.add("invalid");
+  const el = document.createElement("span");
+  el.classList.add("invalid-message");
+  el.textContent = message;
+  parent.appendChild(el);
+};
+
+const resetInputField = (e) => {
+  const className = ["invalid", "valid"];
+  e.target.parentElement.classList.remove(...className);
+  const errorMessage = e.target.parentElement.querySelector(".invalid-message");
+  errorMessage && errorMessage.remove();
+};
+
+const toggleDisableSubmitButton = () => {
+  const result = Object.keys(constraint).every((key) => {
+    const fieldElement = document.getElementById(key).value;
+    return isBlankInInput(fieldElement) || constraint[key].validation() ? false : true;
+  });
+  submitButton.disabled = result ? false : true;
+};
+
+const togglePasswordButton = (e) => {
+  const target = e.target;
+  if (target.classList.contains("is-hide")) {
+    target.previousElementSibling.type = "text";
+    target.classList.remove("is-hide");
+    target.classList.add("is-show");
+  } else {
+    target.previousElementSibling.type = "password";
+    target.classList.remove("is-show");
+    target.classList.add("is-hide");
+  }
+}
+
+const isTokenParam = (params) => {
+  const tokenParam = params.get("token");
+  const token = "482r22fafah";
+  return tokenParam === token;
+}
+
+const isUserParma = (params) => {
+  const userParam = params.get("user");
+  return userParam === localStorage.username;
+}
+
+const checkUrlParams = () => {
+  const urlParams = new URLSearchParams(window.location.search);
+  if (isTokenParam(urlParams) && isUserParma(urlParams)) {
+    window.location.href = "../register/password.html";
+  } else {
+    window.location.href = "../notauthorize.html";
+  }
+}
+
+window.location.search && checkUrlParams();
+
+passwordButtons.forEach(button => {
+  button.addEventListener("click", togglePasswordButton);
+});
+password.addEventListener("blur", isValidField);
+password.addEventListener("focus", resetInputField);
+confirmPassword.addEventListener("blur", isValidField);
+confirmPassword.addEventListener("focus", resetInputField);
+
+
+submitButton.addEventListener("click", (e) => {
+  localStorage.setItem("password", password.value);
+  window.location.href = "./password-done.html";
+  e.preventDefault();
```

Comment:
Q
これってこの関数の一番上に書いたら壊れますか？

Reply:
It won't break.
I thought it would be easier to follow the process if it was separated from the validation process, so I fixed it.
[c00916d](https://github.com/sae-github/JS_lesson/pull/32/commits/c00916d95334c6feb8f0505d1ba69c2a5cd63f5d)

Reply:
あ、ごめんなさい。関数を上に書くと言うよりは

```js
submitButton.addEventListener("click", (e) => {
  e.preventDefault();
  localStorage.setItem("password", password.value);
  window.location.href = "./password-done.html";
  ```
  こうです

Reply:
I see.
Indeed, this is more natural, considering that the process is handled from the top.

Fixed.
[322156a](https://github.com/sae-github/JS_lesson/pull/32/commits/322156af83c27c20ac08383290f1085f9b369375)